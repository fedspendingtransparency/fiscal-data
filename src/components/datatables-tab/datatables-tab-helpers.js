export const fileSizeTranslator = fileSize => {
  const fileSizeNumber = Number(fileSize);
  if (fileSizeNumber < 1000000) {
    const theNumber = Math.round(fileSizeNumber / 1000);
    if (theNumber === 1000) {
      return '1 MB';
    } else {
      return theNumber.toString() + ' KB';
    }
  } else {
    return Math.round(fileSizeNumber / 1000000).toString() + ' MB';
  }
};

/*
 *             ~~~~~~~fileSizeTranslator2~~~~~~~
 *                   This time it's binary
 */
export const fileSizeTranslator2 = fileSize => {
  if (fileSize && !isNaN(fileSize)) {
    const fileSizeNumber = Number(fileSize);
    const fileSizes = ['B', 'KB', 'MB', 'GB'];
    const kb = 1024;
    let curMultiplier = 1;
    for (let i = 0, il = fileSizes.length; i < il; i++) {
      const curThreshold = curMultiplier * kb;
      if (curThreshold > fileSizeNumber) {
        return Math.round(fileSizeNumber / curMultiplier) + ` ${fileSizes[i]}`;
      }
      curMultiplier = curThreshold;
    }
  }

  return null;
};

export const makeTheDataArray = apisArray => {
  const theDataArray = [];
  apisArray.forEach(api => {
    // todo - DTG-2209 will provide the file sizes.
    const dataObj = {
      name: api.tableName,
      description: api.tableDescription,
      rowCount: api.rowCount,
      rowDescription: api.rowDefinition,
    };
    theDataArray.push(dataObj);
  });
  return theDataArray;
};
