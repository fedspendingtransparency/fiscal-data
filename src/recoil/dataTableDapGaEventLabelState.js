import { atom } from 'recoil';

export const dataTableDapGaEventLabelState = atom({
  key: 'dataTableDapGaEventLabelState',
  default: '',
});
