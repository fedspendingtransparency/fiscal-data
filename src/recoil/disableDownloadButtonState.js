import { atom } from 'recoil';

export const disableDownloadButtonState = atom({
  key: 'disableDownloadButtonState',
  default: false,
});
