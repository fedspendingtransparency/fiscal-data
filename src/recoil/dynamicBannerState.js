import { atom, selector } from "recoil";
import { apiPrefix, basicFetch } from "../utils/api-utils";

export const dynamicBannerLastCachedState = atom({
  key: 'dynamicBannerLastCachedState',
  default: 0,
});

export const dynamicBannerState = selector({
  key: 'dynamicBannerState',
  get: async () => {
    const announcementsEndpoint = 'v1/reference/fiscal_data/announcements';
    const response = await basicFetch(`${apiPrefix}${announcementsEndpoint}`);
    const timeStamp = Date.now();
    return {
      payload: response.data,
      timeStamp: timeStamp,
    };
  },
});
