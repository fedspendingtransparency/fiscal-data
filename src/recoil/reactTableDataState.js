import { atom } from 'recoil';

export const reactTablePageState = atom({
  key: 'reactTablePage',
  default: 0,
});

export const reactTableDePaginatedDataState = atom({
  key: 'reactTableDePaginatedDataState',
  default: null,
});
