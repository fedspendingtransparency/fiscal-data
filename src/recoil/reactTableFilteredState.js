import { atom } from 'recoil';

export const reactTableFilteredDateRangeState = atom({
  key: 'reactTableFilteredDateRangeState',
  default: null,
});

export const reactTableSortingState = atom({
  key: 'reactTableSortingState',
  default: [],
});
