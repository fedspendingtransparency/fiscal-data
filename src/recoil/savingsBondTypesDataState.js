import { selector, atom } from 'recoil';
import { apiPrefix, basicFetch } from '../utils/api-utils';

export const savingsBondTypesLastCachedState = atom({
  key: 'savingsBondTypesLastCachedState ',
  default: 0,
});

export const savingsBondTypesData = selector({
  key: 'savingsBondTypesData',
  get: async ({ get }) => {
    const numOfSavingsBonds = 'v1/accounting/od/securities_sales?filter=security_type_desc:eq:Savings%20Bond';
    const response = await basicFetch(`${apiPrefix}${numOfSavingsBonds}&page[size]=1`).then(metaRes => {
      if (metaRes.meta && typeof metaRes.meta['total-pages'] !== 'undefined') {
        const pageSize = metaRes.meta['total-pages'];
        return basicFetch(`${apiPrefix}${numOfSavingsBonds}&page[size]=${pageSize}`);
      }
    });
    const timeStamp = Date.now();
    return {
      payload: response.data,
      timeStamp: timeStamp,
    };
  },
});
