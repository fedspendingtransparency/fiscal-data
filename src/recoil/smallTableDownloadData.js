import { atom } from 'recoil';

export const smallTableDownloadDataCSV = atom({
  key: 'smallTableDownloadDataCSV',
  default: [],
});

export const smallTableDownloadDataJSON = atom({
  key: 'smallTableDownloadDataJSON',
  default: '',
});

export const smallTableDownloadDataXML = atom({
  key: 'smallTableDownloadDataXML',
  default: {},
});

export const tableRowLengthState = atom({
  key: 'tableRowLengthState',
  default: null,
});
