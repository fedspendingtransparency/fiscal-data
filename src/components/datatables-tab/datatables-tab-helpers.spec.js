import React from 'react';
import { fileSizeTranslator, fileSizeTranslator2, makeTheDataArray } from './datatables-tab-helpers';
import renderer from 'react-test-renderer';
import DataTablesTab from './datatables-tab';
import { RecoilRoot } from 'recoil';

describe('DataTablesTabHelpers', () => {
  const mockData = [
    {
      tableName: 'table 1',
      tableDescription: 'table 1 description',
      rowCount: '9999',
      rowDefinition: 'this row does this',
    },
    {
      tableName: 'table 2',
      tableDescription: 'table 2 description',
      rowCount: '111',
      rowDefinition: 'that row does that',
    },
  ];

  let component = renderer.create();
  let instance;
  beforeAll(() => {
    renderer.act(() => {
      component = renderer.create(
        <RecoilRoot>
          <DataTablesTab apis={mockData} />
        </RecoilRoot>
      );
    });
    instance = component.root;
  });

  it('has a function called makeTheDataArray that grabs properties from the objects in the apis array and makes the array to send to the dtgTable component for the datatables tab', () => {
    expect(makeTheDataArray(mockData)).toMatchSnapshot();
  });
  it('has a fileSizeTranslator function that returns file sizes in KB or MB, as strings', () => {
    expect(fileSizeTranslator('999999.999')).toBe('1 MB');
    expect(fileSizeTranslator('1000000')).toBe('1 MB');
    expect(fileSizeTranslator('2905646.0672')).toBe('3 MB');
  });
  it('has a fileSizeTranslator2 function that returns correct file sizes in KB, MB or GB as strings', () => {
    expect(fileSizeTranslator2('156')).toBe('156 B');
    expect(fileSizeTranslator2('999999.999')).toBe('977 KB');
    expect(fileSizeTranslator2(2500000)).toBe('2 MB');
    expect(fileSizeTranslator2(1073741824)).toBe('1 GB');
  });
  it('has a fileSizeTranslator2 function that returns nulls if inappropriate input variables are passed in', () => {
    expect(fileSizeTranslator2(null)).toBeNull();
    expect(fileSizeTranslator2('hello world')).toBeNull();
    expect(fileSizeTranslator2(undefined)).toBeNull();
    expect(fileSizeTranslator2({})).toBeNull();
    // We should never expect a file size to be 0
    expect(fileSizeTranslator2(0)).toBeNull();
  });
});
