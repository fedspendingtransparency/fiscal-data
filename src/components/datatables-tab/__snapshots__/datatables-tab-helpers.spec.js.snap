// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTablesTabHelpers has a function called makeTheDataArray that grabs properties from the objects in the apis array and makes the array to send to the dtgTable component for the datatables tab 1`] = `
Array [
  Object {
    "description": "table 1 description",
    "name": "table 1",
    "rowCount": "9999",
    "rowDescription": "this row does this",
  },
  Object {
    "description": "table 2 description",
    "name": "table 2",
    "rowCount": "111",
    "rowDescription": "that row does that",
  },
]
`;
